### DISCUSSION TITLE
Device interfaces when using slicing

### DISCUSSION THREAD

**Discussion by naveci** [netbox-community/netbox#15116] [Q&A] [answered]
Dear community,

I've been searching a bit on how to deal with this situation, but i feel like this might be a bit of a niche situation. Hence why I'm opening this discussion here. I have the following situation. Most network devices will have a fixed port configuration, however this is not necessarily the case for all such as with a Cisco ASR-9902 for example. [Here's some documentation](https://www.cisco.com/c/en/us/td/docs/iosxr/asr9000/hardware-install/9001-hig/b-asr9001-hardware-installation-guide/b-asr9001-hardware-installation-guide_chapter_01.html)

It allows the operator to 'slice' the port configuration. This router has two NPUs and when i use the following slicing commands, it will result in the port numbers below:
```
hw-module location 0/0/CPU0 slice 0 config-mode 1x100GE,1x100GE,10x10GE,10x10GE
hw-module location 0/0/CPU0 slice 1 config-mode 1x100GE,1x100GE,1x100GE,1x100GE
```

interfaces as a result:
```
interface TenGigE0/0/0/0
interface TenGigE0/0/0/1
interface TenGigE0/0/0/2
interface TenGigE0/0/0/3
interface TenGigE0/0/0/4
interface TenGigE0/0/0/5
interface TenGigE0/0/0/6
interface TenGigE0/0/0/7
interface TenGigE0/0/0/12
interface TenGigE0/0/0/13
interface TenGigE0/0/0/14
interface TenGigE0/0/0/15
interface TenGigE0/0/0/16
interface TenGigE0/0/0/17
interface TenGigE0/0/0/18
interface TenGigE0/0/0/19
interface TenGigE0/0/0/20
interface TenGigE0/0/0/21
interface TenGigE0/0/0/22
interface TenGigE0/0/0/23
interface HundredGigE0/0/0/10
interface HundredGigE0/0/0/11
interface HundredGigE0/0/0/36
interface HundredGigE0/0/0/37
interface HundredGigE0/0/0/38
interface HundredGigE0/0/0/39
```

The bottom four HundredGigE interfaces are part of slice 1, whereas the rest is a result of the config of slice0. There are a whole bunch of variations possible.
Now the question is, how to best define the interfaces/device_types in netbox? Making a new device_type with the slicing information in netbox seems like a lot of work. Also, when an operator would change the slicing, i guess the device would have to be recreated with a different device_type. I'm also not hundred percent sure if this can be solved with interfaces the way it's recommended elsewhere when splitting 40Gbe into 4x10 for example.
Yes, there's a limited number of physical ports, but the numbers differs really on this slicing.

Thanks for your advice.

**Comment by candlerb** [comment, marked as answer]
I had something similar with Nexus C36180YC switches, where a QSFP28 can be either 1 x 40G/100G or 4 x 10G/25G. They call this [breakout](https://www.cisco.com/c/en/us/td/docs/switches/datacenter/nexus9000/sw/7-x/interfaces/configuration/guide/b_Cisco_Nexus_9000_Series_NX-OS_Interfaces_Configuration_Guide_7x/b_Cisco_Nexus_9000_Series_NX-OS_Interfaces_Configuration_Guide_7x_chapter_010.html#concept_855836F7B4C24198891FC6A5FACF6DE7), and it can either be configured globally, or port-by-port ("dynamic breakout").

For example, in the config, Ethernet1/49 vanishes, and is replaced with Ethernet1/49/1, Ethernet1/49/2, Ethernet1/49/3, Ethernet1/49/4.

So in Netbox, I just did exactly the same: delete the QSFP28 interface, and add four SFP28 interfaces. I did this at the level of the device instance, not the device type. (It's perfectly fine to add and remove interfaces on a device instance; the device gets a copy of the interface templates at the time it is instantiated, and any subsequent changes to the device type do not affect any existing device instances)

This meant that my breakout cables appeared as four cables, instead of one cable. I could instead have kept the Ethernet1/49, and created Ethernet1/49/1..4 as virtual interfaces with Ethernet1/49 as their parent. This models the physical breakout cable more accurately (one termination at one side and 4 terminations at the other), but it loses the ability to trace from each individual interface to its peer.

EDIT: It looks like your ASR might be slightly different. As far as I can see, there's a fixed set of physical ports, but depending on how you configure the slicing, some ports may be active or inactive. You could model that by putting all the physical ports in the device type, and then when you change the slicing config, mark the unused ports as "disabled".  (Or you could remove them, but then it's harder to put them back if you change your mind on the config)

**Reply by naveci** [reply]
Thanks for your feedback. It's indeed not exactly like a breakout, but I'm guessing also not too far from it. 
I'll discuss this to see what makes sense in our situation. I was kind of hoping to use that 'disabled' check-box as an admin shut checkbox for automation purposes. So maybe i'd go with the initial proposal. Just need to quickly figure out how many actually do deviate from the default slicing to see how scalable a per-device deviation is 😅

Thanks :)

**Reply by candlerb** [reply]
Cool. If you want to use "enabled" for admin shutdown, then there are other ways you could mark unused interfaces (tags, custom field, label, description). And if you don't change the slicing often (which I doubt you will) then simply deleting them is not so bad. You can recreate them from the interface templates on the device type using [this script](https://github.com/netbox-community/customizations/blob/master/scripts/add_device_type_components.py) - or write your own custom script to handle slicing.
